src/contract/method.rs

```diff
@@ -131,19 +131,22 @@ impl<F> MethodFuture<F> {
     fn new(function: Function, inner: F) -> Self {
         MethodFuture { function, inner }
     }
+
+    /// Get a pinned projection to the inner future.
+    fn inner(self: Pin<&mut Self>) -> Pin<&mut F> {
+        unsafe { self.map_unchecked_mut(|f| &mut f.inner) }
```

Comment:
Oh my, do you really want to add unsafe code to your library? Would it be possible to wrap inner into a `Pin<Box<F>>` or use a library like [pin_project](https://docs.rs/pin-project/0.4.6/pin_project/index.html)?

Reply:
Here be dragons... This is what pin_project generates for us anyway.

Reply:
I can add some comments as to why this is ok to do.

Reply:
I think there is still a difference if you are using a library that is using unsafe code or if you are writing it yourself. E.g. at some point someone might go ahead and declare MethodFuture as Unpin, which as far as I understand would be unsafe. However with Pin<Box> that would be ok and pin_project has some other smart guarantees (e.g. that it doesn't let you incorrectly override the destructor).

Generally I'd defer to well tested and widely adopted libraries when relying on unsafe code and not encourage it to be used directly in our library.

Reply:
`impl Unpin for MethodFuture` is incorrect but never unsafe in itself, but it would require `unsafe` in order to pin the inner future for polling. That unsafe code would be wrong and potentially cause undefined behaviour.

That is a fair point, the smart guarantees of `pin_project` are definitely worth it to not have to roll these unsafe projections by hand.